src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2CryptographicGenerator.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.Git.CredentialManager.Authentication.OAuth
+{
+    public enum OAuth2PkceChallengeMethod
+    {
+        Plain = 0,
+        Sha256 = 1
+    }
+
+    public interface IOAuth2CodeGenerator
+    {
+        /// <summary>
+        /// Create a random string value suitable for use as a nonce.
+        /// </summary>
+        /// <returns>A random string.</returns>
+        string CreateNonce();
+
+        /// <summary>
+        /// Create a cryptographically random code verifier for use with Proof Key for Code Exchange (PKCE).
+        /// </summary>
+        /// <returns>PKCE code verifier.</returns>
+        string CreatePkceCodeVerifier();
+
+        /// <summary>
+        /// Create a code challenge for the given Proof Key for Code Exchange (PKCE) code verifier.
+        /// </summary>
+        /// <param name="challengeMethod">Challenge method.</param>
+        /// /// <param name="codeVerifier">PKCE code verifier.</param>
+        /// <returns>PKCE code challenge.</returns>
+        string CreatePkceCodeChallenge(OAuth2PkceChallengeMethod challengeMethod, string codeVerifier);
+    }
+
+    public class OAuth2CryptographicCodeGenerator : IOAuth2CodeGenerator
+    {
+        // Do not include padding of the base64url string to avoid percent-encoding when passed in a URI
+        private const bool PkceIncludeBase64UrlPadding = false;
+
+        public string CreateNonce()
+        {
+            return Guid.NewGuid().ToString("N");
+        }
+
+        public string CreatePkceCodeVerifier()
+        {
+            //
+            // RFC 7636 requires the code verifier to match the following ABNF:
+            //
+            //   code-verifier = 43*128unreserved
+            //   unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
+            //   ALPHA = %x41-5A / %x61-7A
+            //   DIGIT = %x30-39
+            //
+            // The base64url encoding (RFC 6749) is a subset of these characters
+            // so we opt to convert our random bytes into Base64URL format for
+            // the code verifier.
+            //
+            // RFC 7636 mandates the code verifier must be between 43 and 128 characters
+            // long (inclusive). We want to generate a string at the top end of this range
+            // for maximum entropy. At the same time we want to avoid using the padding
+            // character '=' because this character is percent-encoded when used in URLs.
+            // To avoid padding we need the number of input bytes to be divisible by 3.
+            //
+            // In order to achieve 128 base64url characters AND avoid padding we should
+            // generate exactly 96 random bytes. Why 96 bytes? 96 is divisible by 3 and:
+            //
+            //   96 bytes -> 768 bits -> 128 base64url characters (6 bits per character)
+            //
+            var buf = new byte[96];
```

Comment:
Yes, this is also how we do it https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/94e6e2e3baf840b9e06c8afbe2197d8ec0c0871f/src/client/Microsoft.Identity.Client/Platforms/netcore/NetCoreCryptographyManager.cs#L21



But we use 32 bytes... and perform the [padding logic](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/94e6e2e3baf840b9e06c8afbe2197d8ec0c0871f/src/client/Microsoft.Identity.Client/Utils/Base64UrlHelpers.cs#L69) - base64url encoding (which is classical base64 + padding) is very common in OAuth ...

Reply:
Is there a reason for the 32 bytes? (42.6667 base64 chars with a pad that'll always be trimmed)?

Reply:
I don't remember the reason, there was probably a reference implementation somewhere that was used. It did pass security reviews though.

Reply:
But obviously your argument of using 96 bytes for more entropy is valid, this is a better implementation.

Reply:
I'll log a bug on MSAL.

Reply:
The nice thing about 96 bytes is this works out to a base64url string that doens't need padding!
We still trim any padding characters to be doubly sure however 😁